File: lib/plugins/opabsl/jsbsl/bslClosure.js
/*
    Copyright © 2011, 2012 MLstate

    This file is part of Opa.

    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*/

/**
 * This module provides primitives to manipulates Opa Closure.
 * @side{both}
 * @externType Closure.t
 * @externType Closure.args
 * @opaType Closure.info
 *
 * @author Sebastien Briais
 * @author David Rajchenbach-Teller (made the code more JS idiomatic)
 *
 * @author Quentin Bourgerie
 */

if(IS_OPA_CLIENT) global = this;

//////////////////////////////////////////////////
// Closure arguments
//////////////////////////////////////////////////

/** @module Args */

  /**
   * @register {int -> Closure.args} create
   * @pure
   */
  function closure_args_create(n) {
      return new Array(n);
  }

  /**
   * @register {Closure.args -> int} length
   */
  function closure_args_length(a) {
      return a.length;
  }

  /**
   * @register {Closure.args, int, 'a -> void} set
   */
  function closure_args_set(a,i,x) {
      a[i] = x;
  }

  /**
   * @register {Closure.args, int -> 'a} get
   */
  function closure_args_get(a,i) {
      return a[i];
  }
/** @endModule */

//////////////////////////////////////////////////
// Closure Information
//////////////////////////////////////////////////

/**
 * Retrieve a closure from an identifier.
 *
 * @register {string -> opa[option(Closure.t)]}
 */
function get_local(str) {
  var f = global[str];
  return (f == null) ? js_none : js_some(f);
}

/**
 * Indicates closure presence/absence on the other side from its identifier.
 *
 * @register {string -> bool}
 */
function on_distant(str) {
  var f = global[str];
  return (typeof f != 'undefined') && f.distant || false
}

/**
 * FIXME: right now we have to rename all bypasses to avoid clashes
 * in the global namespace. However, set_distant is used directly by
 * the Opa client code builder. Therefore, we can't rename this function,
 * which can be done right now by creating the binding explicitely.
 *
 * @register {string, bool -> void} set_distant set_distant
 * @opacapi
 */
function set_distant(str, b) {
  var l = str.split(",")
  for (var i = l.length; i--;) {
    (l[i] in global) && (global[l[i]].distant = b)
  }
}

function get_closure_name(closure) {
    var tmp;
    // opa - see type Closure.info
    tmp = closure.info ;
    if (typeof tmp != 'undefined') return tmp.closure_name;

    // chrome
    tmp = closure.name;
    if (tmp) return tmp;

    // all
    return closure.toString().match(/function *([^(]*)/)[1];
}

/**
 * @register {Closure.t -> opa[option(Closure.info)]}
 */
function get_info(closure) {
    var info = closure.info;
    if (typeof info != 'undefined') {
       return js_some(info);
    } else {
      var name = get_closure_name(closure);
      return (global[name] == closure) ? js_some(make_onefield_record(static_field_of_name("closure_name"),name)) : js_none;
    }
}

/**
 * @register {Closure.t, Closure.info -> void}
 */
function set_info(closure, info) {
    closure.info = info;
}

/**
 * Get arguments stored in the closure.
 * @see : env_apply_with_ty
 *
 * @register {Closure.t -> Closure.args}
 */
function get_args(closure) {
    var res = closure.opa_args;
    return (res?res:[]);
}

/**
 * Get type of arguments stored in the closure.
 * @see : env_apply_with_ty
 *
 * @register {Closure.t -> Closure.args}
 */
function get_ty_args(closure) {
    var res = closure.opa_ty_args;
    return (res?res:[]);
}

/**
 * Returns the emptyness of the closure.
 * As we use JavaScript closures, only toplevel function are empty closures.
 *
 * @register {Closure.t -> bool} is_empty
 */
function closure_is_empty(closure) {
    var name = get_closure_name(closure)
    return (global[name] == closure) //
}

//////////////////////////////////////////////////
// Closure
//////////////////////////////////////////////////

/**
 * Create an Opa closure
 * @param [f] an native implementation
 * @param [n] the arity of the closure
 * @param [info] the information of the closure
 *
 * @register {'impl, int, Closure.info -> Closure.t}
 * @pure
 */
function create_anyarray(f, n, info) {
  var new_closure = function() { return f.call(null,arguments) };
  new_closure.info = info;
  return new_closure;
}

/**
 * Cps version of [create_anyarray].
 *
 * @register {'impl, int, Closure.info, continuation(Closure.t) -> void}
 * @pure
 * @cpsBypass
 */
function create_anyarray_cps(f, n, info, k) {
    var any_cps = function(){
        var args = Array.prototype.slice.call(arguments);
        var k = args.pop();
        return f(args, k);
    }
    any_cps.info = info;
    return_(k, any_cps);
}

/**
 * Get from an Opa closure to the native implementation.
 * Note : Identity because Opa use JavaScript Closure.
 */
/**
 * @register {Closure.t -> 'a} export
 * @pure
 */
function export_(clos) {
  return clos;
}

/**
 * Apply arguments [args] to [closure].
 *
 * @register {Closure.t, Closure.args -> 'a}
 */
function apply(closure,args) {
  return closure.apply(null,args)
}

/**
 * Cps version of [apply]
 *
 * @register {Closure.t, Closure.args, continuation('a) -> void}
 * @cpsBypass
 */
function apply_cps(closure, args, k) {
    // Warning : partial apply was broken
    args.push(k);
    var r = BslClosure_apply(closure, args);
    if (r && r !== js_void) push(task_from_return(r[0], [r[1]]));
}

/**
 * Part of JsInterface (funaction)
 */
function _env_apply_with_ty(closure,args,ty_args)
{
   // if env is empty, we should not apply, since it would triger the function computation
   // we only want to apply env, not to do any computation
   // in zero arity no extra lambda is waiting this empty env
   if( args.length ==0 ){
      return closure;
    } else {
      var new_closure           = closure.apply(null,args);
      new_closure.info          = option2js(BslClosure_get_info(closure));
      new_closure.opa_args      = args;
      new_closure.opa_ty_args   = ty_args;
      return new_closure;
    }
}
/** @register {Closure.t, Closure.args,Closure.args -> Closure.t} env_apply_with_ty  _env_apply_with_ty */

/**
 * This function is used by the runtime renaming of client js ast, it's goal is
 * to have same identifer on client and server side.
 * Moreover set the field 'distant' of the closure to true, indeed if the
 * closure is renamed, it means they are a client implementation.
 *
 * @register {string, string -> void}
 */
function replace_identifier(prev, next) {
    var tmp = global[prev];
    // [tmp == null] if it's a function on client side only.
    if(tmp != null&&tmp.constructor==Function){
      // We known only renamed closure will be on the client side, all other
      // closure are just not present on client side or be cleaned.
      tmp.distant  = true;
      // Force the new name of closure (see get_closure_name) this construction
      // have the highest priority
      tmp.info = {closure_name : next};
      global[next] = tmp;
    }
    return js_void;
}

File: lib/plugins/opabsl/mlbsl/bslClosure.ml
(*
    Copyright © 2011, 2012 MLstate

    This file is part of Opa.

    Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*)
##extern-type Closure.args = QmlClosureRuntime.AnyArray.t
##extern-type Closure.t = QmlClosureRuntime.t
##opa-type Closure.info
##module Args
  ##register create \ `QmlClosureRuntime.AnyArray.create` : int -> Closure.args
  ##register length \ `QmlClosureRuntime.AnyArray.length` : Closure.args -> int
  ##register set \ `QmlClosureRuntime.AnyArray.set` : Closure.args, int, 'a -> void
  ##register get \ `QmlClosureRuntime.AnyArray.get` : Closure.args, int -> 'a
##endmodule

##register show \ `QmlClosureRuntime.show` : Closure.t -> string
##register create \ `QmlClosureRuntime.create` : 'impl, int, Closure.info -> Closure.t
##register create_no_function \ `QmlClosureRuntime.create_no_function` : int, Closure.info -> Closure.t
##register [opacapi] define_function \ `QmlClosureRuntime.define_function` : Closure.t, 'impl -> void
##register apply \ `QmlClosureRuntime.args_apply` : Closure.t, Closure.args -> 'a
##register env_apply_with_ty \ `QmlClosureRuntime.env_apply_with_ty` : Closure.t, Closure.args,Closure.args -> Closure.t

##register is_empty \ `QmlClosureRuntime.is_empty` : Closure.t -> bool
##register get_info \ `QmlClosureRuntime.get_identifier` : 'closure -> option(Closure.info)
##register set_info \ `QmlClosureRuntime.set_identifier` : Closure.t, Closure.info -> void
##register get_args \ `QmlClosureRuntime.get_args` : Closure.t -> Closure.args
##register get_ty_args \ `QmlClosureRuntime.get_tyargs` : Closure.t -> Closure.args
##register export \ `QmlClosureRuntime.export` : Closure.t -> 'a
##register import \ `QmlClosureRuntime.import` : 'a, int -> Closure.t

(** Create an "anyarray" closure. These closures consider that the
    given native implementation is a function that takes an anyarray
    of size n rather than a function that take n arguments.
    see also [opavalue.opa (OpaValue.Closure)]
*)
##register create_anyarray \ `QmlClosureRuntime.create_anyarray` : 'impl, int, Closure.info -> Closure.t

(** Create an "anyarray" closure with cps code transformation.
    i.e. We suppose that the given impl is a native function
    typed by [Closure.args -> 'a cont -> unit].
    see also [opavalue.opa (OpaValue.Closure)]
*)
##register [cps-bypass] create_anyarray_cps : 'impl, int, Closure.info, continuation(Closure.t) -> void
let create_anyarray_cps func arity ident k =
  let any_cps oargs =
    let args = QmlClosureRuntime.AnyArray.sub2 oargs 0 arity in
    let k = QmlClosureRuntime.AnyArray.get oargs arity in
    ((Obj.magic func) : _ (*args*) -> _ (*cont*) -> _) args k
  in
  (* arity+1 for continuation added by cps *)
  let cl = QmlClosureRuntime.create_anyarray any_cps (arity + 1) ident in
  QmlCpsServerLib.return k cl

##register [cps-bypass] apply_cps : Closure.t, Closure.args, continuation('a) -> void
let apply_cps closure args k =
  (* Add k only on full application *)
  assert (
    if (
      closure.QmlClosureRuntime.arity - 1 =
        (QmlClosureRuntime.AnyArray.length closure.QmlClosureRuntime.args)
        + (QmlClosureRuntime.AnyArray.length args)
    ) then true else (
      Printf.printf "BslClosure.apply_cps:\n  closure:%s\n  args:%s\n  k:%s\n"
        (DebugPrint.print closure)
        (DebugPrint.print args)
        (DebugPrint.print k);
      false
    )
  );
  let args = QmlClosureRuntime.AnyArray.append args [|Obj.repr k|] in
  QmlClosureRuntime.args_apply closure args


(** A function used to print closure-related debug info*)
##register closure_debug_from_opa: string, string -> void
let closure_debug_from_opa __topic __msg =
  #<If:CLOSURE_DEBUG>
    Logger.debug "[CLOSURE][%s] %s %!" __topic __msg
  #<Else>
    ()
  #<End>



(* CLOSURE REGISTERING *)
(** Type of the information. *)
type t = {
  mutable local : QmlClosureRuntime.t;
  mutable distant : bool;
}

(** The association table. *)
let funtbl : (string, t) Hashtbl.t = Hashtbl.create 1024

let set str t = Hashtbl.add funtbl str t


##register on_distant : string -> bool
(** [on_distant "toto"] If returns true the "toto" function is
    present on the other side. *)
let on_distant str =
  Hashtbl.mem funtbl str && (Hashtbl.find funtbl str).distant

let fclosure_name = ServerLib.static_field_of_name "closure_name"

let make_closure_name name =
  ServerLib.make_record (ServerLib.add_field ServerLib.empty_record_constructor fclosure_name (ServerLib.wrap_string name))

##register [opacapi] create_and_register : 'impl, int, string, bool -> Closure.t
let create_and_register impl arity name distant =
  let ident = make_closure_name name in
  let closure = QmlClosureRuntime.create impl arity ident in
  set name { local = closure; distant = distant };
  closure

##register [opacapi] create_no_function_and_register : int, string, bool -> Closure.t
let create_no_function_and_register arity name distant =
  let ident = make_closure_name name in
  let closure = QmlClosureRuntime.create_no_function arity ident in
  set name { local = closure; distant = distant };
  closure

##register get_local : string -> option(Closure.t)
(** Get the local function, if returns [None] functions is not
    present locally. *)
let get_local str =
  try
    Some (Hashtbl.find funtbl str).local
  with Not_found -> None

##register [opacapi] set_distant : string, bool -> void
(** If a client function get cleaned, set its distant property to
    false.  No need to create a container if the function was not
    already known. *)
let set_distant str b =
  try
    (Hashtbl.find funtbl str).distant <- b;
  with Not_found ->
    ()

##register replace_identifier: string, string -> void
let replace_identifier key_ident ident =
  try
    let {distant; local=clos} = Hashtbl.find funtbl key_ident in
    assert (clos.QmlClosureRuntime.identifier <> None);
    clos.QmlClosureRuntime.identifier <- Some (Obj.repr (make_closure_name ident));
    Hashtbl.remove funtbl key_ident;
    Hashtbl.add funtbl ident {distant; local=clos};
   #<If:JS_RENAMING>
     Printf.printf "Closure update: %s to %s\n%!" key_ident ident;
   #<End>
  with Not_found ->
    #<If:JS_RENAMING>
      Printf.printf "Closure NO update: %s to %s\n%!" key_ident ident;
    #<End>;
    ()
